pkg/domain/domain.go

```diff
@@ -371,7 +373,17 @@ func (do *Domain) loadInfoSchema(startTS uint64, isSnapshot bool) (infoschema.In
 	}
 	infoschema_metrics.LoadSchemaDurationLoadAll.Observe(time.Since(startTime).Seconds())
 
-	builder := infoschema.NewBuilder(do, do.sysFacHack, do.infoCache.Data, useV2)
+	data := do.infoCache.Data
+	if isSnapshot {
+		// Use a NewData() to avoid adding the snapshot schema to the infoschema history.
+		// Why? imagine that the current schema version is [103 104 105 ...]
+		// Then a snapshot read require infoschem version 53, and it's added
+		// Now the history becomes [53,  ... 103, 104, 105 ...]
+		// Then if a query ask for version 74, we'll mistakenly use 53!
```

Comment:
And now the query for 74 will fail? Is this expected?

Reply:
Of course not.
Not we get two kinds of  infoschema, with non-snapshot one [ ... 103, 104, 105 ...]
and snapshot infoschema contains no history like [53]    [74] etc